### PR TITLE
Pruthvi/rnn fusion

### PR DESCRIPTION
**Pull request by pruthviIntel**
- fuses independent inputs across time steps which shares the same weights. this should improve RNN GEMM performance

**Before fusion pass**
[Function_0_000_N6ngraph4pass14NopEliminationE.pdf](https://github.com/NervanaSystems/ngraph/files/2412417/Function_0_000_N6ngraph4pass14NopEliminationE.pdf)

**After fusion pass**
[Function_0_004_N6ngraph7runtime3cpu4pass18CPURnnMatFusion_v2E.pdf](https://github.com/NervanaSystems/ngraph/files/2412424/Function_0_004_N6ngraph7runtime3cpu4pass18CPURnnMatFusion_v2E.pdf)
[Function_0_007_N6ngraph7runtime3cpu4pass9CPUFusionE.pdf](https://github.com/NervanaSystems/ngraph/files/2412425/Function_0_007_N6ngraph7runtime3cpu4pass9CPUFusionE.pdf)

### CODE REVIEW

**Comment by pruthviIntel** [comment]
@jbobba I tried to make changes to existing graph pass `CPURnnMatFusion` to fit both use cases but it becomes more hacky to make it work with the existing code.  a separate graph pass looks lot cleaner IMO, let me know your feedback.

File: src/ngraph/runtime/cpu/pass/cpu_mat_fusion.cpp [hunk 1]
```diff
@@ -102,28 +115,61 @@ bool runtime::cpu::pass::CPURnnMatFusion::run_on_function(std::shared_ptr<Functi
...
         std::make_shared<pattern::Matcher>(bias_pattern)};
 
+    // this DS will be used to hold the matched attributes from matchers_v1
     std::map<std::shared_ptr<Node>, NodeVector> op_seg_map; // add to list of params
     std::map<NodeVector, NodeVector> param_list;
+    //--------------------------------------------------------
+
+    //--------------------------------------------------------
+    // Construct pattern version_2 for RNN input linear transformation
+    auto input_data_pred = [](std::shared_ptr<Node> n) {
```
Comment (Krovatkin) [inline src/ngraph/runtime/cpu/pass/cpu_mat_fusion.cpp]: cud u please use `has_class`?

File: src/ngraph/runtime/cpu/pass/cpu_mat_fusion.cpp [hunk 2]
```diff
@@ -152,54 +198,122 @@ bool runtime::cpu::pass::CPURnnMatFusion::run_on_function(std::shared_ptr<Functi
...
-
-            // replace old nodes
-            function->replace_node(op, slice_node);
+            auto weights = it.first.first;
+            auto bias = it.first.second;
+
+            if (map_weights_to_pattern[weights].size() !=
+                map_weights_bias_to_data[std::make_pair(weights, bias)].size())
+            {
+                throw ngraph_error(
```
Comment (Krovatkin) [inline src/ngraph/runtime/cpu/pass/cpu_mat_fusion.cpp]: this isn't really an error. It's just a case you aren't going to handle

File: src/ngraph/runtime/cpu/pass/cpu_mat_fusion.cpp [hunk 3]
```diff
@@ -152,54 +198,122 @@ bool runtime::cpu::pass::CPURnnMatFusion::run_on_function(std::shared_ptr<Functi
...
+            // now concat the parameter hashed to the same weights
+            auto concated_data = std::make_shared<op::Concat>(data_param_nodes, 0);
+
+            auto& data_shape = concated_data->get_shape();
+            auto data_order = ngraph::get_default_order(concated_data->get_shape());
+            auto& w_shape = weights->get_shape();
+
+            if (w_shape.size() != 2)
+            {
+                throw ngraph_error("weights shape for linear transformation of input must be 2D");
```
Comment (Krovatkin) [inline src/ngraph/runtime/cpu/pass/cpu_mat_fusion.cpp]: ditto

Comment (Krovatkin) [reply src/ngraph/runtime/cpu/pass/cpu_mat_fusion.cpp]: also might wanna move it before you construct a Concat node?

File: src/ngraph/runtime/cpu/pass/cpu_mat_fusion.cpp [hunk 4]
```diff
@@ -152,54 +198,122 @@ bool runtime::cpu::pass::CPURnnMatFusion::run_on_function(std::shared_ptr<Functi
...
+            auto new_input_node = data_shape.size() == 2 ? concated_data : input_reshape_node;
+            auto w_reshape_node = std::make_shared<op::Reshape>(
+                weights, AxisVector{1, 0}, Shape{w_shape[1], w_shape[0]});
+            auto new_dot = std::make_shared<op::Dot>(new_input_node, w_reshape_node);
+            auto bias_broadcast_node =
+                std::make_shared<op::Broadcast>(bias, new_dot->get_shape(), AxisSet{0});
+            auto new_add_bias = std::make_shared<op::Add>(new_dot, bias_broadcast_node);
+
+            // now slice the new_add and feed the corrsponding root nodes
+            auto batch_size = new_add_bias->get_shape()[0] / data_param_nodes.size();
```
Comment (Krovatkin) [inline src/ngraph/runtime/cpu/pass/cpu_mat_fusion.cpp]: is batch_size always going to be the same? Should we have a check for that?

File: src/ngraph/runtime/cpu/pass/cpu_mat_fusion.cpp [hunk 5]
```diff
@@ -152,54 +198,122 @@ bool runtime::cpu::pass::CPURnnMatFusion::run_on_function(std::shared_ptr<Functi
...
+            auto shape_axis_1 = new_add_bias->get_shape()[1];
+            size_t start_index = 0;
+            size_t end_index = batch_size;
+            for (auto& matched_root_node : map_weights_to_pattern[weights])
+            {
```
Comment (Krovatkin) [inline src/ngraph/runtime/cpu/pass/cpu_mat_fusion.cpp]: you could probably just use `start_index` and `batch_size`?

File: src/ngraph/runtime/cpu/pass/cpu_mat_fusion.cpp [hunk 6]
```diff
@@ -152,54 +198,122 @@ bool runtime::cpu::pass::CPURnnMatFusion::run_on_function(std::shared_ptr<Functi
...
+            auto& data_shape = concated_data->get_shape();
+            auto data_order = ngraph::get_default_order(concated_data->get_shape());
+            auto& w_shape = weights->get_shape();
+
+            if (w_shape.size() != 2)
+            {
+                throw ngraph_error("weights shape for linear transformation of input must be 2D");
+            }
+
+            // insert rehape on the concated data to make it 2D, if its 3D
```
Comment (jbobba) [inline src/ngraph/runtime/cpu/pass/cpu_mat_fusion.cpp]: reshape.

**Comment by Krovatkin** [comment]
>  I assume Nik went through the patterns in more detail.
Yes, I did.

**Pull request merged**